xarray/testing/strategies.py

```diff
@@ -0,0 +1,761 @@
+        Can either be a strategy for generating a list of string dimension names,
+        or a strategy for generating a mapping of string dimension names to integer lengths along each dimension.
+        If provided in the former form the lengths of the returned Variable will either be determined from the
+        data argument if given or arbitrarily generated if not.
+        Default is to generate arbitrary dimension names for each axis in data.
+    attrs: Strategy which generates dicts, optional
+    convert: Callable
+        Function which accepts one numpy array and returns one numpy-like array of the same shape.
+        Applied to the data after it is drawn from the `data` strategy provided.
+        Useful for converting numpy arrays to other types of arrays, e.g. sparse arrays.
+        Default is a no-op.
+    """
+
+    if any(
+        not isinstance(arg, st.SearchStrategy) and arg is not None
+        for arg in [data, dims, attrs]
+    ):
+        raise TypeError(
+            "Contents must be provided as a hypothesis.strategies.SearchStrategy object (or None)."
+            "To specify fixed contents, use hypothesis.strategies.just()."
+        )
+
+    if data is not None and dims is None:
+        # no dims -> generate dims to match data
+        data = draw(data)
+        dims = draw(dimension_names(min_ndims=data.ndim, max_ndims=data.ndim))
+
+    elif dims is not None and data is None:
+        # no data -> generate data to match dims
+        dims = draw(dims)
+        if isinstance(dims, List):
+            valid_shapes = npst.array_shapes(min_dims=len(dims), max_dims=len(dims))
+            data = draw(np_arrays(shape=draw(valid_shapes)))
+        else:
+            # should be a mapping of form {dim_names: lengths}
+            shape = tuple(dims.values())
+            data = draw(np_arrays(shape=shape))
+
+    elif data is not None and dims is not None:
+        # both data and dims provided -> check drawn examples are compatible
+        dims = draw(dims)
+
+        # TODO is there another way to enforce these assumptions? This is very like to fail hypothesis' health checks
+        # TODO how do I write a test that checks that the hypothesis Unsatisfiable error will be raised?
+        # TODO or we could just raise in this case?
+        if isinstance(dims, List):
+            data = draw(data)
+            assume(data.ndim == len(dims))
+        else:
+            # should be a mapping of form {dim_names: lengths}
+            data = draw(data)
+            shape = tuple(dims.values())
+            assume(data.shape == shape)
+
+    else:
+        # nothing provided, so generate everything consistently by drawing dims to match data
+        data = draw(np_arrays())
+        dims = draw(dimension_names(min_ndims=data.ndim, max_ndims=data.ndim))
+
+    if isinstance(attrs, st.SearchStrategy):
+        attrs = draw(attrs)
+    elif attrs is None:
+        # TODO autogenerate some attributes
```

Comment:
I left attrs for later because I thought we might want to test storing recursive dicts though... (Also because attrs is like the least important part of this PR)

Reply:
Note: `st.from_type(Hashable)` can generate `Decimal("snan")`, which raises an exception when hashed.

Internally to Hypothesis we have a `.filter(can_hash_without_error)` in a few places to deal with this.